text/global-fixpoints.md

```diff
@@ -0,0 +1,161 @@
+- Title: Named fixpoints and cofixpoints
+
+- Drivers: Hugo Herbelin
+
+----
+
+# Summary and motivation
+
+The Calculus of Inductive Constructions (CIC) originally comes with anonymous fixpoints of the form `fix f args := body(f)` and cofixpoints of the form `cofix f args := body(f)` with reduction rules which - when allowed to be triggered - lead respectively to `body(fix f args := body(f))` and `body(cofix f args := body(f))`.
+
+It is however common to want to hide the size or complexity of such expression behind a name.
+
+This can easily be done with a definition, and this is the way it is done since quite the beginning of the CIC (Coq version 5). Hiding a fixpoint or cofixpoint expression behind a name has however some drawbacks with respect to reduction:
+
+- reduction, such as defined by the CIC, discards the original name and a "refolding" step, such as `simpl` or `cbn` do is necessary to express again the fixpoint or cofixpoint in term of its name
+- moreover, this reduction involves quite a number of steps, see an example below (*)
+
+The objective of this Coq enhancement proposal is to provide "natively" named fixpoints and cofixpoints whose reduction rules directly produce what we intuitively expect, i.e. in, the case (*) below, directly go from `add (S t) m` to `S (add t) m`.
+
+The idea is well-known and used in other proof assistants (e.g. Adga): it is that the body of the fixpoint or cofixpoint refers to the global name associated to the fixpoint rather than to the root of the anonymous `fix` or `cofix` expression.
```

Comment:
FWIW: while this change sounds good, AFAIK Agda (not Adga) goes further, like Equations: the only reduction rules are for `add (S m) n -> S (add m) n` and `add 0 n -> n`. Anything else is neutral (including AFAIK `add` itself; even their analogue of the `match` construct is derived by lambda-lifting to the top-level). That might not be flexible enough for Coq, but in this common case reduction Just Works (tm) out of the box. But I've just used it, not investigated the relevant metatheory.

Reply:
But to be sure — this comment probably amounts to "here's a possible research direction, which I'm not fully qualified to judge".

Reply:
The basic question is whether you want to have a new world with a new `add` in parallel to the current `add` or if you want to extend the current `add`. In general, I tend to be on the side of preserving the existing with minimal changes, so that we can continue to benefit of all what has been done so far.

PS: typo fixed, thanks.